### PR TITLE
Update shard newShardEntry to respect NoFinalize and take ident.Tags or ident.TagsIterator

### PR DESCRIPTION
**Pull request by robskillington**

### CODE REVIEW

File: src/dbnode/storage/shard.go [hunk 1]
```diff
@@ -955,38 +988,57 @@ func (s *dbShard) tryRetrieveWritableSeries(id ident.ID) (
...
 			return nil, err
 		}
 
-		if err := convert.ValidateMetric(clonedID, tags); err != nil {
+		if err := convert.ValidateMetric(newID, tags); err != nil {
 			return nil, err
 		}
+
+	case tagsArg:
+		tags = tagsArgOpts.tags
```
Comment (richardartoul) [inline src/dbnode/storage/shard.go]: this is not pooled / its safe to do this right?

File: src/dbnode/storage/shard.go [hunk 2]
```diff
@@ -1675,31 +1727,23 @@ func (s *dbShard) Bootstrap(
...
-			tagsIter.Reset(dbBlocks.Tags)
-			entry, err = s.insertSeriesSync(dbBlocks.ID, tagsIter,
+			entry, err = s.insertSeriesSync(dbBlocks.ID, newTagsArg(dbBlocks.Tags),
 				insertSyncIncReaderWriterCount)
 			if err != nil {
 				multiErr = multiErr.Add(err)
 				continue
 			}
+		} else {
+			// No longer as we found the series and we don't require
```
Comment (richardartoul) [inline src/dbnode/storage/shard.go]: no longer needed*

File: src/dbnode/storage/shard.go [hunk 3]
```diff
@@ -95,6 +97,38 @@ const (
...
 
+type tagsArgType uint
+
+const (
+	tagsInvalidArg tagsArgType = iota
+	tagsIterArg
+	tagsArg
+)
+
+type tagsArgOptions struct {
```
[the remark this answers is not in the document]
Comment (robskillington) [reply src/dbnode/storage/shard.go]: Sure thing, sounds good.

File: src/dbnode/storage/shard.go [hunk 4]
```diff
@@ -955,38 +988,57 @@ func (s *dbShard) tryRetrieveWritableSeries(id ident.ID) (
...
 	// finalized.
 	// Since series are purged so infrequently the overhead of not releasing
 	// back an ID to a pool is amortized over a long period of time.
-	clonedID := ident.BytesID(append([]byte(nil), id.Bytes()...))
-	clonedID.NoFinalize()
+	var newID ident.BytesID
+	if id.IsNoFinalize() {
+		// If the ID is already marked as NoFinalize, meaning it won't be returned
+		// to any pools, then we can directly take reference to it.
+		// We make sure to use ident.BytesID for this ID to avoid inc/decref when
```
[the remark this answers is not in the document]
Comment (robskillington) [reply src/dbnode/storage/shard.go]: Good catch, thanks.

File: src/dbnode/storage/shard.go (hunk 1 above)
Comment (robskillington) [reply src/dbnode/storage/shard.go]: Yeah, in the case where we actually pass through an `ident.Tags` it is not pooled and is returned from bootstrappers.

**Pull request merged**